pod-overhead: add Overhead to PodSpec and RuntimeClass

egernst opened the pull request:
**What type of PR is this?**
 /kind feature

**What this PR does / why we need it**:

Addition of overhead field to the PodSpec as part of the API changes described in the [Pod Overhead KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md)

**Does this PR introduce a user-facing change?**:

Yes 

 /sig node


```release-note
Addition of Overhead field to the PodSpec and RuntimeClass types as part of the Pod Overhead KEP
```

Commit:
pod-overhead: add Overhead to PodSpec internal type

Update internal PodSpec to make use of Overhead field. Add validation
and validation tests.

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

Commit:
pod-overhead: add Overhead to PodSpec

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

Commit:
pod-overhead: add Overhead to RuntimeClass internal type

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

Commit:
pod-overhead: add Overhead to RuntimeClass

Co-authored-by: Tim Allclair <tallclair@google.com>
Signed-off-by: Eric Ernst <eric.ernst@intel.com>

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2973,6 +2973,14 @@ type PodSpec struct {
 	// +optional
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// Overhead represents the overhead associated with running a pod for a given runtimeClass. This field will be
+	// populated at admission time by the RtunimeClass admission controller. Any prior values being set will result
```
```suggestion
	// autopopulated at admission time by the RuntimeClass admission controller. Any prior values being set will result
```

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2973,6 +2973,14 @@ type PodSpec struct {
+	// in the workload being rejected. If runtimeClass is configured and selected in the PodSpec, Overhead will be set
+	// to the value defined in the corresponding runtimeClass.
+	// More info: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md
+	// This is an alpha feature and may change in the future.
```
```suggestion
	// This field is alpha-level as of Kubernetes v1.15, and is only honored by servers that enable the PodOverhead feature.
```

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2973,6 +2973,14 @@ type PodSpec struct {
 	// Optional: Defaults to true.
 	// +optional
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// Overhead represents the overhead associated with running a pod for a given runtimeClass. This field will be
```
```suggestion
	// Overhead represents the resource overhead associated with running a pod for a given RuntimeClass. This field will be
```

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2973,6 +2973,14 @@ type PodSpec struct {
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// Overhead represents the overhead associated with running a pod for a given runtimeClass. This field will be
+	// populated at admission time by the RtunimeClass admission controller. Any prior values being set will result
+	// in the workload being rejected. If runtimeClass is configured and selected in the PodSpec, Overhead will be set
```
Maybe this wording is clearer?
```diff
- Any prior values being set will result in the workload being rejected.
+ Overhead must not be set in Pod create requests, or the request will be rejected.
```

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2973,6 +2973,14 @@ type PodSpec struct {
+	// Overhead represents the overhead associated with running a pod for a given runtimeClass. This field will be
+	// populated at admission time by the RtunimeClass admission controller. Any prior values being set will result
+	// in the workload being rejected. If runtimeClass is configured and selected in the PodSpec, Overhead will be set
+	// to the value defined in the corresponding runtimeClass.
```
nit: be consistent with RuntimeClass capitalization.

tallclair replied:
```suggestion
	// to the value defined in the corresponding runtimeClass, otherwise it will remain unset and treated as zero.
```

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2973,6 +2973,14 @@ type PodSpec struct {
+	// populated at admission time by the RtunimeClass admission controller. Any prior values being set will result
+	// in the workload being rejected. If runtimeClass is configured and selected in the PodSpec, Overhead will be set
+	// to the value defined in the corresponding runtimeClass.
+	// More info: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190226-pod-overhead.md
```
nit: prefer
```suggestion
	// More info: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md
```

tallclair reviewed (commented):
some wording tweaks

tallclair commented:
/priority important-soon
/sig node
/milestone v1.15

egernst commented:
Thanks for the quick feedback @tallclair - updated, PTAL!

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -2675,6 +2675,15 @@ type PodSpec struct {
+	// More info: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md
+	// This field is alpha-level as of Kubernetes v1.15, and is only honored by servers that enable the PodOverhead feature.
+	// +optional
+	// +optional
```
duplicate line

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -2675,6 +2675,15 @@ type PodSpec struct {
 	// This is an alpha feature and may change in the future.
 	// +optional
 	RuntimeClassName *string
+	// Overhead represents the resource  overhead associated with running a pod for a given RuntimeClass. This field will be
```
nit: extra space
```suggestion
	// Overhead represents the resource overhead associated with running a pod for a given RuntimeClass. This field will be
```

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3018,6 +3018,10 @@ func ValidatePodSpec(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
 
+	if spec.Overhead != nil {
+		allErrs = append(allErrs, field.Forbidden(fldPath.Child("overhead"), "may only be set during admission"))
+	}
```
Hmm, I'm not sure whether validation runs before or after admission... my guess would be after since we would want to validate admission mutations...

Also, this should be ignored if the PodOverhead feature is disabled.

I think it might make more sense to just enforce this in the admission controller. That way we can also ignore the error if it is set to the same value as the RuntimeClass.

egernst replied:
Makes sense.  Addressed.

tallclair reviewed (commented):
If the PodOverhead feature is not enabled, the overhead field needs to be dropped. See https://github.com/kubernetes/kubernetes/blob/master/pkg/api/pod/util.go#L283

I also think it makes sense to add the RuntimeClass API in the same PR, since those fields are intrinsically linked.

Commit:
pod-overhead: Introduce PodOverhead feature gate

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

yastij commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3018,6 +3018,10 @@ func ValidatePodSpec(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
 		allErrs = append(allErrs, ValidateRuntimeClassName(*spec.RuntimeClassName, fldPath.Child("runtimeClassName"))...)
 	}
 
+	if spec.Overhead != nil {
```
I agree this should be done at the runtimeClass admission controller level

egernst replied:
Definitely makes sense.  Added some basic tests in its place (behind the feature gate).

yastij reviewed (commented):
@egernst - can you add the runtimeClass API on a separate commit ?

Commit:
pod-overhead: drop from PodSpec based on feature-gate

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

egernst commented:
Updated to add the Feature Gate, and remove the original test - that was a misunderstanding on how validation is used.

I added RuntimeClass API changes as well.  Right now I'm not "scrubbing" and dropping Overhead from this (like is done with PodSpec).

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -264,6 +264,33 @@ func ValidateRuntimeClassName(name string, fldPath *field.Path) field.ErrorList
+	limPath := fldPath.Child("limits")
+	reqPath := fldPath.Child("requests")
+
+	//Overhead validation is a superset of Resource Requirements
```
nit: space after //, always

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -264,6 +264,33 @@ func ValidateRuntimeClassName(name string, fldPath *field.Path) field.ErrorList
+	//Overhead validation is a superset of Resource Requirements
+	allErrs = append(allErrs, ValidateResourceRequirements(overhead, fldPath)...)
+
+	//If an overhead for a resource is being defined, both limits and request should be set
```
What does this mean for best-effort pods? Does it even make sense to have separate requests and limits for overhead?

I would like the QoS to be preserved, independent of RuntimeClass (assuming the runtime supports all QoS - I'm not sure if BestEffort makes sense for Kata). I can't remember if the KEP covered this?

egernst replied:
> What does this mean for best-effort pods? Does it even make sense to have separate requests and limits for overhead?

I think it makes sense to have seperate still ('burstable').  As this PR is written right now, I was under the mindset that leaving 'limits' unset was a bad idea.  Ie, even for a burstable Overhead, there should be a limit defined. I no longer think this makes sense; leaving it able to scale unbound on host could be useful, and it doesn't make sense to change the existing pattern on how ResourceRequirements are used today.

> I would like the QoS to be preserved, independent of RuntimeClass (assuming the runtime supports all QoS - I'm not sure if BestEffort makes sense for Kata). I can't remember if the KEP covered this?

For a workload, unbound burstable (limits not set) are not well support with Kata (ie, additional vCPUs are added based on a limits value provided).

egernst replied:
TL;DR  -- I'll update validation/default logic to match what is done today for core.ResourceRequirements .

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -264,6 +264,33 @@ func ValidateRuntimeClassName(name string, fldPath *field.Path) field.ErrorList
+	allErrs = append(allErrs, ValidateResourceRequirements(overhead, fldPath)...)
+
+	//If an overhead for a resource is being defined, both limits and request should be set
+	for overheadName, _ := range overhead.Requests {
```
nit: omit `, _`

tallclair commented on pkg/apis/core/validation/validation.go:
```diff
@@ -264,6 +264,33 @@ func ValidateRuntimeClassName(name string, fldPath *field.Path) field.ErrorList
+	//If an overhead for a resource is being defined, both limits and request should be set
+	for overheadName, _ := range overhead.Requests {
+		_, exists := overhead.Limits[overheadName]
+		if !exists {
```
nit: inline
```suggestion
		if _, exists := overhead.Limits[overheadName]; !exists {
```

tallclair commented on staging/src/k8s.io/api/node/v1beta1/types.go:
```diff
@@ -48,6 +49,19 @@ type RuntimeClass struct {
+
+// Overhead structure represents the resource overhead associated with running a pod.
+type Overhead struct {
+	//PodFixed represents the fixed resource overhead associated with running a pod.
```
space after //

tallclair replied:
Should this field be optional? It doesn't make sense to omit it now, but if we add other fields under overhead I think it should be.

tallclair commented on staging/src/k8s.io/api/node/v1beta1/types.go:
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package v1beta1
```
You also need to add it to the alpha API, I think

tallclair replied:
not positive though... ask sig-apimachinery?

yastij replied:
@tallclair - I agree, I think we should. We have to be able to convert from internal types to versioned without data loss:

1 - v1beta1 -> internal type works
2 - internal type -> v1alpha1 doesn't work if we don't have `Overhead` field

tallclair commented on pkg/apis/node/v1beta1/defaults.go:
```diff
@@ -0,0 +1,51 @@
+	"k8s.io/api/node/v1beta1"
+)
+
+func SetDefaults_Overhead(obj *v1beta1.Overhead) {
```
Validation and Defaulting need to be registered. I can help you set it up next week, but first I think we should discuss whether requests and limits both make sense for overhead.

saad-ali reviewed (commented):
Meeting Notes From the API review meeting today with @thockin

* Applied by AdmissionController?
  * Yes.
* Previous topic, why make scheduler distinct? In theory you could have secondary node selector (capability set) that you reify from RuntimeClass on the way in, and the scheduler knows its there so even if RuntimeClass is deleted, it will still exist on Pod.
* Call out version skew for `Overhead` -- if scheduler doesn't understand it. Scheduler move forward without API Server and vice versa. Probably ok, but document it.
* Does it go into pod cgroup?
  * Some runtimes do it right. ContainerD puts it on KubeReserved.
  * We need to separate requests and limits. CPU requests are factored in. Kubelet is aware of distinction, doesn't make distinction in memory yet, but does for CPU.
  * I can't guarantee that it will always be accounted in to cgroup. At least scheduler does? Ya, but that's worse, you don't overcommit node but grant pod more memory then it should have. Workload it self will be scheduled with extra buffer.
* That has to be something runtime decides -- where does that allocation go, pod or parent level group?
* Now that CRI has interface for metrics, should kubelet bother creating the cgroups? Kubelet does lifecycle via CRI -- acts as a mux or kubelet acts on multiple CRIs. Inclined to say yes, but it will probably come back and bite us.
  * Accounting for kernel or kubelet overhead (which may use more with every new runtime), I could see two sets of overhead one accounted to pod and a system wide one, but that can be an additive change.
* Called `PodFixed` someday `PodProportional`? 
  * Yes.
* Combinatoric result of 6 resources? 
  * Yes, but computed at admission time so it is one number. In the future it may be more? Yes.
* `Requests` and `Limits` in `PodFixed`? 
  * Yes. Big debate about this. There could be a baseline for pod, and an upper limit. You can turn non burstable in to burstable? No. QoS class is computed excluding overhead. Maybe guaranteed pod over head request to overhead limit. David and Dawn disagreed.

tallclair commented:
In the follow-up discussion, we agreed on collapsing overhead requests & limits into a single value. KEP update here: https://github.com/kubernetes/enhancements/pull/1057

egernst commented:
@tallclair - I've made the changes, but am seeing an issue with the updated/autogenerated code.  AFAICT, an import isn't being created.  I see the following error after doing `make clean && make generated_files && make update`:
```
Running update-generated-docs
# k8s.io/kubernetes/vendor/k8s.io/api/node/v1alpha1
vendor/k8s.io/api/node/v1alpha1/generated.pb.go:426:17: undefined: k8s_io_apimachinery_pkg_api_resource
vendor/k8s.io/api/node/v1alpha1/generated.pb.go:497:18: undefined: k8s_io_apimachinery_pkg_api_resource
# k8s.io/kubernetes/vendor/k8s.io/api/node/v1beta1
vendor/k8s.io/api/node/v1beta1/generated.pb.go:376:17: undefined: k8s_io_apimachinery_pkg_api_resource
vendor/k8s.io/api/node/v1beta1/generated.pb.go:447:18: undefined: k8s_io_apimachinery_pkg_api_resource
!!! [0520 22:31:53] Call tree:
!!! [0520 22:31:53]  1: /home/eernst/go/src/k8s.io/kubernetes/hack/lib/golang.sh:715 kube::golang::build_some_binaries(...)
!!! [0520 22:31:53]  2: /home/eernst/go/src/k8s.io/kubernetes/hack/lib/golang.sh:854 kube::golang::build_binaries_for_platform(...)
!!! [0520 22:31:53]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0520 22:31:53] Call tree:
!!! [0520 22:31:53]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0520 22:31:53] Call tree:
!!! [0520 22:31:53]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [all] Error 1
Running update-generated-docs FAILED
```

tallclair commented:
@egernst, I pushed a fix. There's a "magic" protobuf hint that I pulled from the [metrics api](https://github.com/kubernetes/kubernetes/blob/3517197bcb34b9702a54647ce694cddb2446362e/staging/src/k8s.io/metrics/pkg/apis/metrics/v1alpha1/types.go#L41) for the resource list, which fixed the problem. I also reverted the conversion function change you made, since those functions are already there in the generated conversion files. The commits on this PR are getting a bit unwieldy, so you might want to squash them & rebase on HEAD. LMK if you want me to take care of that.

egernst commented:
Thanks for the help @tallclair.  PR rebased, and squashed fixes into original commits.

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2982,6 +2982,14 @@ type PodSpec struct {
 	EnableServiceLinks *bool `json:"enableServiceLinks,omitempty" protobuf:"varint,30,opt,name=enableServiceLinks"`
+	// Overhead represents the resource overhead associated with running a pod for a given RuntimeClass. This field will be
+	// autopopulated at admission time by the RuntimeClass admission controller. Overhead must not be set in Pod create requests,
+	// or the request will be rejected. If RuntimeClass is configured and selected in the PodSpec, Overhead will be set
```
Who rejects?  Is it the admission controller (I think it is?) or apiserver?  If it is admission controller we should explain that here, and detail that it CAN be disabled (which makes this comment wrong as-is).

egernst replied:
Indeed, the admission controller would reject.  You're right that it can be disabled - I'll update the comments to clarify this.

thockin commented on pkg/apis/node/validation/validation.go:
```diff
@@ -41,3 +49,10 @@ func ValidateRuntimeClassUpdate(new, old *node.RuntimeClass) field.ErrorList {
 }
+
+func validateOverhead(overhead *node.Overhead, fldPath *field.Path) field.ErrorList {
+	resourceReq := core.ResourceRequirements{
```
Short comment on the trickiness here would be nice.

egernst replied:
addressed

thockin commented on pkg/apis/node/v1alpha1/conversion.go:
```diff
@@ -33,11 +33,23 @@ func addConversionFuncs(s *runtime.Scheme) error {
 	out.ObjectMeta = in.ObjectMeta
 	out.Handler = in.Spec.RuntimeHandler
+	if in.Spec.Overhead != nil {
+		out.Overhead = &node.Overhead{}
```
Why do we need it to be non-nil when converted?

tallclair replied:
Just the way the auto-generated conversion function is written, it needs a non-nil object to update. Note that we only do this if the input overhead is non-nil.

thockin reviewed (commented):
Mostly requests for comments and clarity.

egernst commented:
Noted I still need to do one more rebase —onto ; will do later this afternoon.

tallclair commented:
Fixed golint failure, I had the fix locally but it never got pushed.

xmudrii commented:
Hello! We have started the code freeze for 1.15. Is this PR still planned for the 1.15 cycle?

egernst commented:
@xmudrii I'm hopeful this lands -- waiting on ACK from approvers still

egernst commented:
@vishh, @thockin  - can you PTAL?

thockin reviewed (commented):
API approved

/approve

thockin commented:
I wanted @tallclair to LGTM, but he seems MIA.

In his stead (and I hope he agrees)

/lgtm

egernst commented:
The e2e perf looked like noise. Can we retest?

liggitt commented:
fyi, ran this against the [in-progress API compatibility tests](https://github.com/kubernetes/kubernetes/pull/78309) and the changes in this PR do not affect round-tripping of 1.14.x-level data, so 👍

egernst commented:
@tallclair I’m flying next few hours, can you help rebase? I should be able to fix around noon pst.

liggitt commented:
FYI, we just added tests to verify we don't break compatibility with previously serialized data in #78309. Part of that keeps up-to-date serializations for current API types in `staging/src/k8s.io/api/testdata/HEAD`. That means that PRs that add fields (like this one), need to update those fixtures.

When this is rebased on master, run the following to update the fixtures (because it touches PodSpec, it will update all the fixtures for things that embed PodSpec as well):
```sh
UPDATE_COMPATIBILITY_FIXTURE_DATA=true go test ./vendor/k8s.io/api -run //HEAD
```

xmudrii commented:
Hello! The code freeze has started last week. As this PR seems to have a lot of merge conflicts and needs some more work, considering we're already late in the cycle, I would propose moving it for the 1.16 release cycle. If there are any objections, please let the release team know.

/milestone v1.16

egernst commented:
Hey @xmudrii - Yes,  this makes sense.  The implementation in scheduler and a new admission controller were not reviewed in time, so this was no longer a good candidate for 1.15.

Thanks for the help.

Commit:
pod-overhead: autogenerated code updates

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

tallclair commented on pkg/apis/core/types.go:
```diff
@@ -2700,6 +2700,16 @@ type PodSpec struct {
+	// set. If RuntimeClass is configured and selected in the PodSpec, Overhead will be set to the value
+	// defined in the corresponding RuntimeClass, otherwise it will remain unset and treated as zero.
+	// More info: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md
+	// This field is alpha-level as of Kubernetes v1.15, and is only honored by servers that enable the PodOverhead feature.
```
s/1.15/1.16/

tallclair commented on pkg/features/kube_features.go:
```diff
@@ -479,6 +479,12 @@ const (
 	VolumePVCDataSource featuregate.Feature = "VolumePVCDataSource"
+
+	// owner: @egernst
+	// alpha: v1.15
```
1.16

tallclair commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -3001,6 +3001,16 @@ type PodSpec struct {
+	// set. If RuntimeClass is configured and selected in the PodSpec, Overhead will be set to the value
+	// defined in the corresponding RuntimeClass, otherwise it will remain unset and treated as zero.
+	// More info: https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md
+	// This field is alpha-level as of Kubernetes v1.15, and is only honored by servers that enable the PodOverhead feature.
```
1.16

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -45,6 +46,19 @@ type RuntimeClass struct {
+
+	// Overhead represents the resource overhead associated with running a pod for a
+	// given RuntimeClass. For more details, see
+	// https://git.k8s.io/enhancements/keps/sig-node/20190226-pod-overhead.md
```
This should have a comment about being an alpha field, the same as the one on the pod field.

egernst commented on pkg/api/pod/util.go:
```diff
@@ -375,6 +375,11 @@ func dropDisabledFields(
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) && !overheadInUse(oldPodSpec) {
+		// Set Overhead to nil only if the feature is disabled and it is not used
+		podSpec.Overhead = nil
+	}
```
Makes sense, though i didn't see how that relates to storage?

Commit:
pod overhead: drop from RuntimeClass base on feature-gate

Signed-off-by: Eric Ernst <eric.ernst@intel.com>

egernst commented:
This needs a lgtm one more time, since the latest rebase.  ping @thockin

egernst commented:
Prior API review approval for reference: https://github.com/kubernetes/kubernetes/pull/76968#pullrequestreview-244072982

/cc @liggitt @tallclair @thockin @yastij

egernst commented:
/test pull-kubernetes-e2e-gce-100-performance

egernst commented:
/test pull-kubernetes-bazel-test

tallclair commented:
/lgtm

thockin commented on pkg/registry/node/runtimeclass/strategy.go:
```diff
@@ -56,7 +58,12 @@ func (strategy) AllowCreateOnUpdate() bool {
-	_ = obj.(*node.RuntimeClass)
+	rc := obj.(*node.RuntimeClass)
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) && rc != nil {
```
Also need PrepareForUpdate() ?

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3095,6 +3101,10 @@ func ValidatePodSpec(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
 		allErrs = append(allErrs, ValidatePreemptionPolicy(spec.PreemptionPolicy, fldPath.Child("preemptionPolicy"))...)
 	}
 
+	if spec.Overhead != nil && utilfeature.DefaultFeatureGate.Enabled(features.PodOverhead) {
```
If you have the strategy blanking out the fields, you should not even need to check the gate here - it will be nil, right?

thockin reviewed (commented):
API still LGTM
